File: app/src/main/java/com/steven/moviesearch/ResultActivity.java
package com.steven.moviesearch;

import android.app.LoaderManager;
import android.content.Intent;
import android.content.Loader;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.steven.moviesearch.models.ResultItem;

import java.util.List;

public class ResultActivity extends AppCompatActivity implements LoaderManager
		.LoaderCallbacks<List<ResultItem>> {

	private static final String TAG = ResultActivity.class.getName();

	private TextView resultTextView;
	private ResultItemAdapter resultItemAdapter;
	private RecyclerView recyclerView;
	private ProgressBar progressBar;
	private String searchText;
	private Toast toast;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_result);
		if (getIntent().hasExtra("searchText")) {
			searchText = getIntent().getStringExtra("searchText");
		}
		resultItemAdapter = new ResultItemAdapter(this);
		resultTextView = (TextView) findViewById(R.id.resultTextView);
		recyclerView = (RecyclerView) findViewById(R.id.resultRecyclerView);
		progressBar = (ProgressBar) findViewById(R.id.progress_bar);

		resultItemAdapter.setOnItemClickListener(new ResultItemAdapter.OnItemClickListener() {
			@Override
			public void onItemClick(int position) {
				if (Utils.isNetworkAvailable(ResultActivity.this)) {
					Intent i = new Intent(ResultActivity.this, DetailsActivity.class);
					i.putExtra("resultId", resultItemAdapter.getItem(position).getId());
					startActivity(i);
				} else {
					toast = Toast.makeText(ResultActivity.this, "Network Not Available", Toast
							.LENGTH_SHORT);
					toast.show();
				}
			}
		});
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
		recyclerView.setLayoutManager(linearLayoutManager);
		recyclerView.setAdapter(resultItemAdapter);
		showProgress();
		getLoaderManager().initLoader(1, null, this);
	}

	private void showMessage(String message) {
		recyclerView.setVisibility(View.GONE);
		progressBar.setVisibility(View.GONE);
		resultTextView.setVisibility(View.VISIBLE);
		resultTextView.setText(message);
	}

	private void showProgress() {
		recyclerView.setVisibility(View.GONE);
		resultTextView.setVisibility(View.GONE);
		progressBar.setVisibility(View.VISIBLE);
	}

	private void showResults(List<ResultItem> resultItems) {
		resultTextView.setVisibility(View.GONE);
		progressBar.setVisibility(View.GONE);
		recyclerView.setVisibility(View.VISIBLE);
		resultItemAdapter.swapResults(resultItems);
	}

	@Override
	public Loader<List<ResultItem>> onCreateLoader(int id, Bundle args) {
		return new ResultLoader(this, searchText);
	}

	@Override
	public void onLoadFinished(Loader<List<ResultItem>> loader, List<ResultItem> data) {

		if (data == null) {
			showMessage("Error");
		} else if (data.size() == 0) {
			showMessage("No Result");
		} else {
			showResults(data);
		}
	}

	@Override
	public void onLoaderReset(Loader<List<ResultItem>> loader) {
	}
}

File: app/src/main/java/com/steven/moviesearch/ui/ResultActivity.kt
package com.steven.moviesearch.ui

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.LoaderManager.LoaderCallbacks
import android.support.v4.content.Loader
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.widget.Toast
import com.steven.moviesearch.R
import com.steven.moviesearch.ResultItemAdapter
import com.steven.moviesearch.Utils
import com.steven.moviesearch.loaders.ResultLoader
import com.steven.moviesearch.models.ResultItem
import kotlinx.android.synthetic.main.activity_result.*

class ResultActivity : AppCompatActivity(), LoaderCallbacks<List<ResultItem>?> {

	private val resultItemAdapter: ResultItemAdapter by lazy { ResultItemAdapter(this) }
	private var searchText: String? = null
	private var toast: Toast? = null

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_result)
		if (intent.hasExtra("searchText")) {
			searchText = intent.getStringExtra("searchText")
		}

		resultItemAdapter.onItemClickListener = fun(position: Int) {
			val item = resultItemAdapter.getItem(position) ?: return
			if (Utils.isNetworkAvailable(this@ResultActivity)) {
				val intent = Intent(this@ResultActivity, DetailsActivity::class.java)
				intent.putExtra("resultId", item.id)
				intent.putExtra("resultMediaType", item.mediaType)
				startActivity(intent)
			} else {
				toast?.cancel()
				toast = Toast.makeText(this@ResultActivity, "Network Not Available", Toast
						.LENGTH_SHORT)
				toast?.show()
			}
		}

		val linearLayoutManager = LinearLayoutManager(this)
		resultRecyclerView.layoutManager = linearLayoutManager
		resultRecyclerView.adapter = resultItemAdapter
		showProgress()
		supportLoaderManager.initLoader(1, null, this)
	}

	private fun showMessage(message: String) {
		resultRecyclerView.visibility = View.GONE
		progress_bar.visibility = View.GONE
		resultTextView.visibility = View.VISIBLE
		resultTextView.text = message
	}

	private fun showProgress() {
		resultRecyclerView.visibility = View.GONE
		resultTextView.visibility = View.GONE
		progress_bar.visibility = View.VISIBLE
	}

	private fun showResults(resultItems: List<ResultItem>) {
		resultTextView.visibility = View.GONE
		progress_bar.visibility = View.GONE
		resultRecyclerView.visibility = View.VISIBLE
		resultItemAdapter.swapResults(resultItems)
	}

	override fun onCreateLoader(id: Int, args: Bundle?): Loader<List<ResultItem>?> {
		return ResultLoader(this, searchText!!)
	}

	override fun onLoadFinished(loader: Loader<List<ResultItem>?>, data: List<ResultItem>?) {
		when {
			data == null -> showMessage("Error")
			data.isEmpty() -> showMessage("No Result")
			else -> showResults(data)
		}
	}

	override fun onLoaderReset(loader: Loader<List<ResultItem>?>) {

	}

	companion object {
		private val TAG = ResultActivity::class.java.name
	}
}
